### PR TITLE
Feature/#22053 add privacy provider

### PR DESCRIPTION
**Pull request by jtgoltz**

### CODE REVIEW

File: classes/privacy/provider.php [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    \core_privacy\local\request\plugin\provider,
+    \core_privacy\local\request\core_userlist_provider {
+    /**
+     * Returns meta data about this system.
+     *
+     * @param   collection $collection The initialised item collection to add items to.
+     * @return  collection A listing of user data stored through this system.
+     */
+    public static function get_metadata(collection $collection) : collection {
+        debugging("getMetadata");
```
Comment (fsacha) [inline classes/privacy/provider.php:53]: Remove

File: lang/en/mumie.php [hunk 2]
```diff
@@ -158,4 +158,10 @@
...
 $string['mumie_calendar_duedate_desc'] = 'Submissions after this deadline will not be used for grading';
-$string['mumie_calendar_duedate_extension'] = 'Extended due date: {$a}';
\ No newline at end of file
+$string['mumie_calendar_duedate_extension'] = 'Extended due date: {$a}';
+
+// Privacy.
+$string['privacy:metadata:mod_mumie_duedate_extensions:mumie'] = 'ID of MUMIE Task the extension is for';
+$string['privacy:metadata:mod_mumie_duedate_extensions:duedate'] = 'Timestamp of due date extension';
+$string['privacy:metadata:mod_mumie_duedate_extensions:userid'] = 'User ID to whom the extension is granted';
+$string['privacy:metadata:mod_mumie_duedate_extensions:tableexplanation'] = 'Due extensions';
```
Comment (fsacha) [inline lang/en/mumie.php:167]: Rename 'Due extensions' into 'Due  date extensions'

File: classes/privacy/provider.php [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+     * @param   int $userid The user to search.
+     * @return  contextlist   $contextlist  The contextlist containing the list of contexts used in this plugin.
+     */
+    public static function get_contexts_for_userid(int $userid) : contextlist {
+        global $DB, $CFG;
+        $contextlist = new contextlist();
+        $contextlist->set_component('mod_mumie');
+
+        require_once($CFG->dirroot . '/mod/mumie/classes/mumie_duedate_extension.php');
+        $mumieids = array_map(
```
Comment (fsacha) [inline classes/privacy/provider.php:78]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+
+    public function test_get_contexts_for_userid_with_data() {
+        $this->resetAfterTest(true);
+
+        $course = $this->getDataGenerator()->create_course();
+        $user = $this->getDataGenerator()->create_user();
+        $anotheruser = $this->getDataGenerator()->create_user();
+
+        $mumie = $this->create_test_mumie_cm($course);
+        $anothermumie = $this->create_test_mumie_cm($course);
```
Comment (fsacha) [inline tests/privacy_provider_test.php:57]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+ * @copyright  2017-2021 integral-learning GmbH (https://www.integral-learning.de/)
+ * @author Tobias Goltz (tobias.goltz@integral-learning.de)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mod_mumie_privacy_provider_testcase extends \core_privacy\tests\provider_testcase {
+    public function test_get_contexts_for_userid_no_data() {
+        global $USER;
+        $this->resetAfterTest(true);
+
+        $this->setAdminUser();
```
Comment (fsacha) [inline tests/privacy_provider_test.php:44]: Why do you use admin? Can't we use '$this->getDataGenerator()->create_user()'?

File: tests/privacy_provider_test.php [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+        $mumie = $this->create_test_mumie_cm($course);
+        $anothermumie = $this->create_test_mumie_cm($course);
+
+        $this->add_duedate($mumie, $user);
+
+        $contextlist = provider::get_contexts_for_userid($user->id);
+        $this->assertCount(1, $contextlist);
+
+        // Make sure that there is no data for the control group.
+        $contextlist = provider::get_contexts_for_userid($anotheruser->id);
```
Comment (fsacha) [inline tests/privacy_provider_test.php:65]: Already tested in 'test_get_contexts_for_userid_no_data()'

File: tests/privacy_provider_test.php [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+        $this->assertCount(0, $contextlist);
```
Comment (fsacha) [inline tests/privacy_provider_test.php:66]: Already tested in 'test_get_contexts_for_userid_no_data()'

File: tests/privacy_provider_test.php [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    public function test_get_users_in_context() {
```
Comment (fsacha) [inline tests/privacy_provider_test.php:69]: I would seperate it in different test cases

File: tests/privacy_provider_test.php [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+/**
+ * Privacy provider test class.
+ *
+ * @package mod_mumie
+ * @copyright  2017-2021 integral-learning GmbH (https://www.integral-learning.de/)
+ * @author Tobias Goltz (tobias.goltz@integral-learning.de)
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+class mod_mumie_privacy_provider_testcase extends \core_privacy\tests\provider_testcase {
```
Comment (fsacha) [inline tests/privacy_provider_test.php:39]: I would prefer this signature for the test cases: [https://martinfowler.com/bliki/GivenWhenThen.html](url)

File: classes/privacy/provider.php [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+        if ($context instanceof \context_module && count($userids) > 0) {
+            list($insql, $inparams) = $DB->get_in_or_equal($userids);
+
+            $innserselect = "SELECT duedate.id FROM {mumie_duedate} duedate
+                            JOIN {course_modules} cm ON cm.instance = duedate.mumie
+                            WHERE cm.id = ?
+                            AND duedate.userid $insql
+                            ";
+            $sql = "id IN ($innserselect)";
```
Comment (fsacha) [inline classes/privacy/provider.php:227]: I prefer the upper variant, where everything is in one variable. Anyway, this should be uniform.

File: tests/privacy_provider_test.php [hunk 11]
```diff
@@ -0,0 +1,295 @@
...
+        provider::export_user_data($approvedcontextlist);
+        $writer = writer::with_context(\context_system::instance());
+        $this->assertFalse($writer->has_any_data_in_any_context());
+    }
+
+    public function test_export_user_data() {
+        $this->resetAfterTest(true);
+
+        $course1 = $this->getDataGenerator()->create_course();
+        $course2 = $this->getDataGenerator()->create_course();
```
Comment (fsacha) [inline tests/privacy_provider_test.php:243]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+        $this->assertCount(1, $DB->get_records('mumie_duedate', array('mumie' => $m2->id)));
+
+    }
+
+    public function test_delete_data_for_user() {
+        global $DB;
+        $this->resetAfterTest(true);
+
+        $course1 = $this->getDataGenerator()->create_course();
+        $course2 = $this->getDataGenerator()->create_course();
```
Comment (fsacha) [inline tests/privacy_provider_test.php:189]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+        $user2 = $this->getDataGenerator()->create_user();
+        $user3 = $this->getDataGenerator()->create_user();
+
+        $this->add_duedate($m1, $user1);
+        $this->add_duedate($m2, $user1);
+        $this->add_duedate($m1, $user2);
+        $this->add_duedate($m1, $user3);
+
+        $context1 = context_module::instance($m1->cmid);
+        $context2 = context_module::instance($m2->cmid);
```
Comment (fsacha) [inline tests/privacy_provider_test.php:172]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+            $DB->get_fieldset_select('mumie_duedate', 'userid', 'mumie = ?', [$m1->id])
+        );
+    }
+
+    public function test_delete_data_for_all_users_in_context() {
+        global $DB;
+        $this->resetAfterTest(true);
+
+        $course1 = $this->getDataGenerator()->create_course();
+        $course2 = $this->getDataGenerator()->create_course();
```
Comment (fsacha) [inline tests/privacy_provider_test.php:157]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+        $user2 = $this->getDataGenerator()->create_user();
+        $user3 = $this->getDataGenerator()->create_user();
+
+        $this->add_duedate($m1, $user1);
+        $this->add_duedate($m2, $user1);
+        $this->add_duedate($m1, $user2);
+        $this->add_duedate($m1, $user3);
+
+        $context1 = context_module::instance($m1->cmid);
+        $context2 = context_module::instance($m2->cmid);
```
Comment (fsacha) [inline tests/privacy_provider_test.php:130]: Is this variable needed?

File: tests/privacy_provider_test.php [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+        provider::get_users_in_context($userlist3);
+        $this->assertEquals([], $userlist3->get_userids());
+    }
+
+    public function test_delete_data_for_users() {
+        global $DB;
+        $this->resetAfterTest(true);
+
+        $course1 = $this->getDataGenerator()->create_course();
+        $course2 = $this->getDataGenerator()->create_course();
```
Comment (fsacha) [inline tests/privacy_provider_test.php:115]: Is this variable needed?

File: tests/privacy_provider_test.php (hunk 7 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php:66]: test_get_contexts_for_userid_no_data works with an empty db. This test is meant to make sure that there are no false positives in a non-empty db.

File: tests/privacy_provider_test.php (hunk 9 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php:39]: 404 - link not working

Comment (jtgoltz) [reply tests/privacy_provider_test.php:39]: Link works, when copied and pasted but not when clicked. Weird.

That naming looks interesting for future test cases, but I don't think we really _need_ to rename all of the privacy tests right now. The current style is taken from core moodle components

File: tests/privacy_provider_test.php (hunk 8 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php:69]: I'm not sure if this is wise in this case. We want to test how a non-empty database behaves in different cases. Of course, we can just run the same code at the beginning to fill it in every single instance of the test - but I don't see how this ends up improving the code.

File: classes/privacy/provider.php (hunk 1 above)
Comment (jtgoltz) [reply classes/privacy/provider.php]: removed

File: lang/en/mumie.php (hunk 2 above)
Comment (jtgoltz) [reply lang/en/mumie.php]: renamed

File: classes/privacy/provider.php (hunk 3 above)
Comment (jtgoltz) [reply classes/privacy/provider.php]: no, it's dead code by now

File: tests/privacy_provider_test.php (hunk 4 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

File: tests/privacy_provider_test.php (hunk 5 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: This test was copied from a core component. I don't see why we would need to use admin for that, so I changed it to your suggestion

File: classes/privacy/provider.php (hunk 10 above)
Comment (jtgoltz) [reply classes/privacy/provider.php]: changed

File: tests/privacy_provider_test.php (hunk 12 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

File: tests/privacy_provider_test.php (hunk 13 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

File: tests/privacy_provider_test.php (hunk 14 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

File: tests/privacy_provider_test.php (hunk 15 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

File: tests/privacy_provider_test.php (hunk 16 above)
Comment (jtgoltz) [reply tests/privacy_provider_test.php]: removed

**Pull request merged**